### ISSUE TITLE
Bug: An Internal Error Has Occurred. - Diagnostics page

### ISSUE THREAD

**Issue by SteveClement** [MISP/MISP#8057]
### Expected behavior

Display diagnostics page.

### Actual behavior

Internal Error

### Steps to reproduce

Open Diagnostics

### Version

2.4.151

### Operating System

Rocky Linux

### Operating System version

8.5

### PHP version

7.4

### Browser

FF

### Browser version

latest

### Relevant log output

```shell
2021-12-25 11:32:00 Error: [Exception] DateTime::__construct(): Failed to parse time string (1637009368) at position 7 (3): Unexpected character
Request URL: /servers/serverSettings/diagnostics
Stack Trace:
#0 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(470): DateTime->__construct()
#1 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(387): SecurityAudit->getKernelBuild()
#2 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(272): SecurityAudit->system()
#3 /var/www/MISP/app/Controller/ServersController.php(1187): SecurityAudit->run()
#4 [internal function]: ServersController->serverSettings()
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs()
#6 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction()
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#8 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
#9 {main}
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by patricksaladino** [comment]
I'm having the same issue. RHEL 8.5, PHP 7.4.19, **MISP version 2.4.152**

Excerpt from /var/www/MISP/app/tmp/logs/error.log :

```
2021-12-27 07:26:41 Error: [Exception] DateTime::__construct(): Failed to parse time string (1639018277) at position 7 (2): Unexpected character
Request URL: /servers/serverSettings/diagnostics
Stack Trace:
#0 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(470): DateTime->__construct('1639018277')
#1 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(387): SecurityAudit->getKernelBuild()
#2 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(272): SecurityAudit->system(Array)
#3 /var/www/MISP/app/Controller/ServersController.php(1187): SecurityAudit->run(Object(Server))
#4 [internal function]: ServersController->serverSettings('diagnostics')
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs(Object(ServersController), Array)
#6 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction(Object(CakeRequest))
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke(Object(ServersController), Object(CakeRequest))
#8 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch(Object(CakeRequest), Object(CakeResponse))
#9 {main}
```

**Comment by righel** [comment]
Hello @patricksaladino, a fix for this issue in https://github.com/MISP/MISP/commit/76a0f0617a89675a6f8330a9b598bc689fe2869f

It may have been pushed after you installed/updated your MISP instance. Could you please update your MISP instance?
Since the UI is not working you could do it via the terminal by running:
`/var/www/MISP/app/Console/cake Admin updateMISP`

Let me know if that helps.

**Comment by patricksaladino** [comment]
Dear @righel,

Thank you for the quick fix. Unfortunately the following commands :

```
sudo -u apache git pull origin 2.4 (ran from the root dir of MISP)
sudo -u apache /var/www/MISP/app/Console/cake Admin updateMISP
```
had no effect as the fix doesn't seem to be integrated (yet) in the main 2.4 branch (`Already up to date.`).

Any recommendations ?

Kind regards,
Patrick

**Comment by righel** [comment]
Have you checked the error is still there when you go to the diagnostics page?
With `sudo -u apache git pull origin 2.4` your MISP code should have been updated to the latest tag and this fix should be included.

**Comment by patricksaladino** [comment]
Error is still there and the code hasn't been updated to the latest revision :-(

![image](https://user-images.githubusercontent.com/17050508/147563475-392fbf97-493d-4c2d-aa2e-4f27f02b7adf.png)
![image](https://user-images.githubusercontent.com/17050508/147563504-063edb41-a3a7-4477-9f7a-e67b5a3de141.png)

**Comment by patricksaladino** [comment]
Interestingly, I'm stuck at rev. ee1ee01c62b88d54ac9b1b5b7613699c785b1164 (3 days ago), while the most recent commit seems to be 76a0f0617a89675a6f8330a9b598bc689fe2869f.

**Comment by righel** [comment]
Could you please share the output of these commands?
```
sudo -u apache git describe --tags
sudo -u apache git rev-parse --short HEAD
```
Last one should show ee1ee01c62b88d54ac9b1b5b7613699c785b1164 as the last rev.

If that's correct the fix should already be there, are you getting the same error on the logs? Maybe it's something else now.

Additionally, could you check your local `/var/www/MISP/app/Lib/Tools/SecurityAudit.php::getKernelBuild()` file looks like this?
https://github.com/MISP/MISP/blob/2.4/app/Lib/Tools/SecurityAudit.php#L459-L474

**Comment by patricksaladino** [comment]
Dear @righel. As asked :

```
[root@misp MISP]# sudo -u apache git describe --tags
v2.4.134-4250-gee1ee01c6

[root@misp MISP]# sudo -u apache git rev-parse --short HEAD
ee1ee01c6
```

I'm still getting the same error (generated about two hours ago) :

```
2021-12-28 12:18:26 Error: [Exception] DateTime::__construct(): Failed to parse time string (1639018277) at position 7 (2): Unexpected character
Request URL: /servers/serverSettings/diagnostics
Stack Trace:
#0 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(470): DateTime->__construct('1639018277')
#1 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(387): SecurityAudit->getKernelBuild()
#2 /var/www/MISP/app/Lib/Tools/SecurityAudit.php(272): SecurityAudit->system(Array)
#3 /var/www/MISP/app/Controller/ServersController.php(1187): SecurityAudit->run(Object(Server))
#4 [internal function]: ServersController->serverSettings('diagnostics')
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs(Object(ServersController), Array)
#6 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction(Object(CakeRequest))
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke(Object(ServersController), Object(CakeRequest))
#8 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch(Object(CakeRequest), Object(CakeResponse))
#9 {main}
```

`/var/www/MISP/app/Lib/Tools/SecurityAudit.php` do integrates the fix deployed with comit https://github.com/MISP/MISP/commit/76a0f0617a89675a6f8330a9b598bc689fe2869f :

![image](https://user-images.githubusercontent.com/17050508/147571354-3fb19df3-9768-4dfa-a705-1b09c4f5d72e.png)

Thank you for your time and efforts !

**Comment by righel** [comment]
This is strange.
https://github.com/MISP/MISP/blob/76a0f0617a89675a6f8330a9b598bc689fe2869f/app/Lib/Tools/SecurityAudit.php#L468
`strtotime()` either returns an `int` or `false`, therefore it should not be possible to `$buildTime` to be of type `string` which seems to be the pre-fix behavior and the error in the logs.

Sanity check: any chance we are looking at code not deployed/deployed to other host?

**Comment by patricksaladino** [comment]
It doesn't make ANY sense...and the PHP snippet below (from /var/www/MISP/app/Lib/Tools/SecurityAudit.php::getKernelBuild()) works as expected :

```
<?php
$version = php_uname('v');
$sub = substr($version, 7);
$buildTime = strtotime($sub);
var_dump ($version);
var_dump ($sub);
var_dump ($buildTime);
?>
```
```
[root@misp ~]# php test.php
string(34) "#1 SMP Wed Dec 8 21:51:17 EST 2021"
string(27) "Wed Dec 8 21:51:17 EST 2021"
int(1639018277)
```
Regarding the sanity check, we only have one production host running MISP. I'd provide you with any needed proof/evidence you'd deem necessary.

**Comment by patricksaladino** [comment]
Quick remark : The DateTime object constructor expects a string (see https://www.php.net/manual/fr/datetime.construct.php) :

`public DateTime::__construct(string $datetime = "now", ?DateTimeZone $timezone = null)`

and we're providing it with an integer (see previous post). Would it be automagically casted/converted to a string ?

**Comment by patricksaladino** [comment]
I finally found the culprit. The UNIX epoch timestamp has to be prefixed with an "@" when used by DateTime::__construct

Working "fix" : 

```
$buildTime = strtotime(substr($version, 7));
        if ($buildTime) {
            return new DateTime('@'.$buildTime);
        } else {
            return false;
        }
```

**Comment by righel** [comment]
~Saw that in the docs, but on my env `new DateTime($buildTime)` works well when $buildTime is of type `int`
Could you share the output of `php_uname('v')` ?~

**Comment by patricksaladino** [comment]
Please see my comment from 41 min ago. Copy/pasted here for your comfort and enjoyment :

```
<?php
$version = php_uname('v');
$sub = substr($version, 7);
$buildTime = strtotime($sub);
var_dump ($version);
var_dump ($sub);
var_dump ($buildTime);
?>
```

```
[root@misp ~]# php test.php
string(34) "#1 SMP Wed Dec 8 21:51:17 EST 2021"
string(27) "Wed Dec 8 21:51:17 EST 2021"
int(1639018277)
```

**Comment by patricksaladino** [comment]
It's confirmed and reproductible. My PHP env. (PHP 7.4.19 (cli) (built: May  4 2021 11:06:37) ( NTS )) definitely needs the ampersand prefix when using an UNIX epoch timestamp with DateTime::__construct. The following code highlights it :

```
<?php
$version = php_uname('v');
$sub = substr($version, 7);
$buildTime = strtotime($sub);
var_dump ($version);
var_dump ($sub);
var_dump ($buildTime);

$dateString =  new DateTime('@'.$buildTime);
var_dump ($dateString);

try
{
	$dateInt = new DateTime($buildTime);
}

catch (Exception $e)
{
	print_r ($e);
}
?>
```
```
[root@misp ~]# php test.php
string(34) "#1 SMP Wed Dec 8 21:51:17 EST 2021"
string(27) "Wed Dec 8 21:51:17 EST 2021"
int(1639018277)
object(DateTime)#1 (3) {
  ["date"]=>
  string(26) "2021-12-09 02:51:17.000000"
  ["timezone_type"]=>
  int(1)
  ["timezone"]=>
  string(6) "+00:00"
}
Exception Object
(
    [message:protected] => DateTime::__construct(): Failed to parse time string (1639018277) at position 7 (2): Unexpected character
    [string:Exception:private] =>
    [code:protected] => 0
    [file:protected] => /root/test.php
    [line:protected] => 14
    [trace:Exception:private] => Array
        (
            [0] => Array
                (
                    [file] => /root/test.php
                    [line] => 14
                    [function] => __construct
                    [class] => DateTime
                    [type] => ->
                    [args] => Array
                        (
                            [0] => 1639018277
                        )

                )

        )

    [previous:Exception:private] =>
    [severity] => 2
)
```

**Comment by righel** [comment]
Thanks for the extensive troubleshooting session @patricksaladino 
Pushed this fix: 40f83a0329e5dde04f45ec8f00c281fd1b2c9c6c

**Comment by patricksaladino** [comment]
Well thank YOU !

Edit : fix is working as expected (commit 40f83a0329e5dde04f45ec8f00c281fd1b2c9c6c)
